activation/nipost_state.go

```diff
@@ -57,58 +59,81 @@ func read(path string) ([]byte, error) {
 	return data[:end], nil
 }
 
-func saveNipostChallenge(dir string, ch *types.NIPostChallenge) error {
-	if _, err := os.Stat(dir); err != nil {
-		return fmt.Errorf("post dir not created: %w", err)
-	}
-	data, err := codec.Encode(ch)
+func load(filename string, dst scale.Decodable) error {
```

Comment:
`load` seems to be inefficient:

1. Read all data into memory
2. Remove last 8 bytes
3. Calculate CRC of data in memory (i.e. has to be copied again)
4. Parse last 8 bytes as Uint64
5. Return all data except last 8 bytes to be decoded
6. Decode data as scalable (i.e. copy data again)

This can be done more efficiently:

1. Move CRC to beginning of file and read first as uint64
2. Use an io.TeeReader to calculate CRC while decoding a the scale encoded data without the need of loading data into memory and copying twice.
3. ???
4. Profit! 🙂

Reply:
The same can be done in reverse `write` already does it mostly effeciently; it uses `io.MultiWriter` to write data to a buffer and CRC at the same time.

I would only change it to not use a buffer, but instead write 8 (zero) bytes to a file then use the `io.MultiWriter` to calculate and write to the file at the same time and then seek to the beginning of the file and update the CRC.

The new file can then atomically replace the old one with `[atomic.ReplaceFile](https://github.com/natefinch/atomic/blob/v1.0.1/file_unix.go#L12)`

Reply:
Good idea :+1:, though it's used infrequently and the data is small so it shouldn't matter.

I just refactored the existing code to reuse it and fixing inefficiencies is out of scope. Could you open an issue for it?

Reply:
Created issue here: https://github.com/spacemeshos/go-spacemesh/issues/4506